record/record_test.go

```diff
@@ -448,7 +497,134 @@ func TestComplexRecover(t *testing.T) {
 		t.Fatalf("Next: %v", err)
 	}
 	r2Data, _ := ioutil.ReadAll(r2)
-	if !bytes.Equal(r2Data, records[2]) {
+	if !bytes.Equal(r2Data, testData.records[2]) {
 		t.Fatal("Unexpected output in r2's data")
 	}
 }
+
+func TestRecoverMultipleBlocks(t *testing.T) {
+	testData, err := makeTestRecords(
+		// The first record will consume 3 entire blocks but a fraction of the 4th.
+		blockSize*3,
+		// The second record will completely fill the remainder of the 4th block.
+		3*(blockSize-headerSize)-2*blockSize-2*headerSize,
+		// Consume the entirety of the 5th block.
+		blockSize-headerSize,
+		// Consume the entirety of the 6th block.
+		blockSize-headerSize,
+		// Consume roughly half of the 7th block.
+		blockSize/2,
+	)
+	if err != nil {
+		t.Fatalf("makeTestRecords: %v", err)
+	}
+
+	// Corrupt the checksum for the portion of the first record that exists in the 4th block.
+	corruptBlock(testData.buf, 3)
+
+	// Now corrupt the two blocks in a row that correspond to testData.records[{2,3}].
+	corruptBlock(testData.buf, 4)
+	corruptBlock(testData.buf, 5)
+
+	// The first record should fail, but only when we read deeper beyond the first block.
+	r := NewReader(bytes.NewReader(testData.buf))
+	r0, err := r.Next()
+	if err != nil {
+		t.Fatalf("Next: %v", err)
+	}
+
+	// Reading deeper should yield a checksum mismatch.
+	_, err = ioutil.ReadAll(r0)
+	if err == nil {
+		t.Fatal("Exptected a checksum mismatch error, got nil")
+	}
+	if !strings.Contains(err.Error(), "checksum mismatch") {
+		t.Fatalf("Unexpected error returned: %v", err)
+	}
+
+	// Recover from that checksum mismatch.
+	r.Recover()
+
+	// All of the data in the second record is lost because the first
+	// record shared a partial block with it. The following two records
+	// have corrupted checksums as well, so the call above to r.Recover
+	// should result in r.Next() being a reader to the 5th record.
+	r4, err := r.Next()
+	if err != nil {
+		t.Fatalf("Next: %v", err)
+	}
+
+	r4Data, _ := ioutil.ReadAll(r4)
+	if !bytes.Equal(r4Data, testData.records[4]) {
+		t.Fatal("Unexpected output in r4's data")
+	}
+}
+
+// verifyLastBlockRecovery reads each record from testData expecting that the
+// last record will be corrupted. It will then try Recover and verify that EOF
+// is returned.
+func verifyLastBlockRecovery(testData *testRecords, t *testing.T) {
+	r := NewReader(bytes.NewReader(testData.buf))
+	// Loop to one element larger than the number of records to verify EOF.
+	for i := 0; i < len(testData.records)+1; i++ {
```

Comment:
This might be easier as:

for i := 0; ; i++ {
  _, err := r.Next()
  switch i {
  default:
    if err != nil {
      etc
    }
  case len(records)-1:
    etc
    continue
  case len(records):
    etc
    return
  }
}

WDYT?